modules/pool/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+pub struct RejectedFragmentRegistration {
+    pub id: FragmentId,
+    pub reason: RegistrationError,
+}
+
+#[derive(Debug, Error)]
+pub enum RegistrationError {
+    #[error(transparent)]
+    Balance(#[from] BalanceError),
+    #[error("the provided fragment must not be present outside of block0")]
+    Block0Fragment,
+    #[error("the provided fragment is currently unsupported")]
+    UnsupportedFragment,
+    #[error("tried to add a fragment already present in the pool")]
+    KnownFragment,
+}
+
+impl FragmentsPool {
+    pub fn new(config: Configuration) -> Self {
+        let pool = lru::LruCache::new(config.max_pool_entries);
+        Self { pool }
+    }
+
+    /// Register new fragments in this pool. If the total number of fragments
+    /// exceeds `Configuration::max_pool_entries`, older pool entries will be
+    /// removed.
+    ///
+    /// # Returns
+    ///
+    /// Information about fragments that cannot be added to the pool.
+    pub fn register(&mut self, fragments: Vec<Fragment>) -> Vec<RejectedFragmentRegistration> {
+        let mut rejected = Vec::new();
+
+        for fragment in fragments {
+            let hash = fragment.hash();
+            if self.pool.peek(&hash).is_some() {
+                rejected.push(RejectedFragmentRegistration {
+                    id: hash,
+                    reason: RegistrationError::KnownFragment,
+                });
+            }
+            if let Err(reason) = verify_fragment(&fragment) {
+                rejected.push(RejectedFragmentRegistration { id: hash, reason });
+            }
+            self.pool.put(hash, fragment);
+        }
+
+        rejected
+    }
+
+    /// Select fragments for a new block. All processed fragments (selected and
+    /// rejected) are removed from the pool.
+    ///
+    /// # Arguments
+    ///
+    /// * `ledger` - current ledger state.
+    /// * `ledger_params` - ledger parameters for the current epoch.
+    /// * `block_date` - the date of the block that is being built.
+    ///
+    /// # Returns
+    ///
+    /// Fragments to be added to the new block and information the ones that
+    /// were rejected.
+    pub fn select(
+        &mut self,
+        mut ledger: Ledger,
+        ledger_params: &LedgerParameters,
+        block_date: BlockDate,
+    ) -> SelectionResult {
+        let fragments_number_max = ledger_params.block_content_max_size;
+
+        let mut selected = Vec::new();
+        let mut rejected = Vec::new();
+
+        while let Some((hash, fragment)) = self.pool.pop_lru() {
+            match ledger.apply_fragment(&ledger_params, &fragment, block_date) {
+                Ok(new_ledger) => {
+                    ledger = new_ledger;
+                    selected.push(fragment);
+                }
+                Err(reason) => rejected.push(RejectedFragmentSelection { id: hash, reason }),
+            }
+
+            if selected.len() as u32 >= fragments_number_max {
```

Comment:
is there a reason this is `>=` instead of `==`?

Reply:
Just a belt-and-suspenders approach :grinning: I also do it instinctively as a safeguard that would prevent the code from continued damage on counter overruns and doesn't make a difference if the functionality is correct.